### ISSUE TITLE
Obtaining a low accuracy with Tiny CNN

### ISSUE THREAD

**Issue by RafaRuiz** [tiny-dnn/tiny-dnn#28]
Hello.

I'm developing a model and I like trying things in MATLAB before, so I trained a model which works with this framework: https://github.com/rasmusbergpalm/DeepLearnToolbox

But when I'm trying to build it with Tiny-CNN, I get an accuracy of around 37% when instead using DeepLearnToolbox I get 98%.

I don't know what I'm doing bad.

Whad I do:
- Create images to ubyte.
- I use the 1st example:

void sample1_convnet(void) {
        // construct LeNet-5 architecture
        network<mse, gradient_descent_levenberg_marquardt> nn;

```
    // connection table [Y.Lecun, 1998 Table.1]
#define O true
#define X false
    static const bool connection [] = {
        O, X, X, X, O, O, O, X, X, O, O, O, O, X, O, O,
        O, O, X, X, X, O, O, O, X, X, O, O, O, O, X, O,
        O, O, O, X, X, X, O, O, O, X, X, O, X, O, O, O,
        X, O, O, O, X, X, O, O, O, O, X, X, O, X, O, O,
        X, X, O, O, O, X, X, O, O, O, O, X, O, O, X, O,
        X, X, X, O, O, O, X, X, O, O, O, O, X, O, O, O
    };
#undef O
#undef X

    nn << convolutional_layer<tan_h>(32, 32, 5, 1, 6) // 32x32 in, 5x5 kernel, 1-6 fmaps conv
       << average_pooling_layer<tan_h>(28, 28, 6, 2) // 28x28 in, 6 fmaps, 2x2 subsampling
       << convolutional_layer<tan_h>(14, 14, 5, 6, 16,
                                     connection_table(connection, 6, 16)) // with connection-table
       << average_pooling_layer<tan_h>(10, 10, 16, 2)
       << convolutional_layer<tan_h>(5, 5, 5, 16, 120)
       << fully_connected_layer<tan_h>(120, 10);

    std::cout << "load models..." << std::endl;

    // load MNIST dataset
    std::vector<label_t> train_labels, test_labels;
    std::vector<vec_t> train_images, test_images;

    parse_mnist_labels("../../data/vuittonoutput_train_labels.idx1", &train_labels);
    parse_mnist_images("../../data/vuittonoutput_train_images.idx3", &train_images);
    parse_mnist_labels("../../data/test_train_labels.idx1", &test_labels);
    parse_mnist_images("../../data/test_train_images.idx3", &test_images);

    std::cout << "start learning" << std::endl;

//    for (int i = 0; i < train_images.size(); i++) {
//        cout << endl << "image " << i << " size = " << train_images.at(i).size();
//    }

    boost::progress_display disp(train_images.size());
    boost::timer t;
    int minibatch_size = 2;

    nn.optimizer().alpha *= std::sqrt(minibatch_size);

    // create callback
    auto on_enumerate_epoch = [&](){
        std::cout << t.elapsed() << "s elapsed." << std::endl;

        tiny_cnn::result res = nn.test(test_images, test_labels);

        std::cout << nn.optimizer().alpha << "," << res.num_success << "/" << res.num_total << std::endl;

        nn.optimizer().alpha *= 0.90; // decay learning rate
        nn.optimizer().alpha = std::max(0.0000001, nn.optimizer().alpha);

        disp.restart(train_images.size());
        t.restart();
    };

    auto on_enumerate_minibatch = [&](){ 
        disp += minibatch_size; 

        // weight visualization in imdebug
        /*static int n = 0;    
        n+=minibatch_size;
        if (n >= 1000) {
            image img;
            C3.weight_to_image(img);
            imdebug("lum b=8 w=%d h=%d %p", img.width(), img.height(), &img.data()[0]);
            n = 0;
        }*/
    };

    // training
    nn.train(train_images, train_labels, minibatch_size, 100, on_enumerate_minibatch, on_enumerate_epoch);

    std::cout << "end training." << std::endl;

    // test and show results
    nn.test(test_images, test_labels).print_detail(std::cout);

    // save networks
    std::ofstream ofs("LeNet-weights");
    ofs << nn;
}
```

am I doing anything bad? I've tried to change some parameters but the result is almost the same.
Thank you very much in advance.

**Comment by RafaRuiz** [comment]
Oh, this is my final accuracy:

accuracy:35.6262% (1155/3242
    *     0     1     2 
    0     0     0     0 
    1     0     0     0 
    2  1118   969  1155

**Comment by wdx04** [comment]
@RafaRuiz 
You can try to use another optimizer(for example: adagrad) and adjust its learning rate(alpha).
Also you should change your last layer to fully_connected_layer<tan_h>(120, 3) since you have 3 classes instead of 10.

**Comment by edgarriba** [comment]
@RafaRuiz Please check if last released version you get those metrics. thx !
https://github.com/nyanp/tiny-cnn/releases/tag/v0.1.0

**Comment by Randl** [comment]
@RafaRuiz  are you still using interested in tiny-dnn?

**Comment by RafaRuiz** [comment]
Hello @Randl I left time ago my job where I was working with that.

Apparently the fault could be in the data set, which was pretty unstable, could that be the reason?

**Comment by Randl** [comment]
Yes it definitely could be a reason

**Comment by taotaoyuhust** [comment]
do you solve this problem
i got the same problem:
*    0    1   2 
0    0    0    0
1    0    0     0
2   271 467 347

**Comment by taotaoyuhust** [comment]
solved 
you should mix the order
